File: agent4/engine/fingerspell.go
package engine

import "strings"

// Fingerspell converts a word into a fingerspelled SignEntry.
// Letters A-Z and digits 0-9 each become one letter in the sequence.
// Duration is ~300ms per letter.
func Fingerspell(word string) SignEntry {
	upper := strings.ToUpper(strings.TrimSpace(word))
	var letters []string
	for _, c := range upper {
		switch {
		case c >= 'A' && c <= 'Z':
			letters = append(letters, string(c))
		case c >= '0' && c <= '9':
			letters = append(letters, string(c))
		}
	}
	if len(letters) == 0 {
		letters = []string{upper}
	}
	return SignEntry{
		Gloss:      upper,
		Type:       SignTypeFingerSpell,
		Letters:    letters,
		DurationMs: len(letters) * 300,
	}
}

File: agent4/asl_engine/fingerspell.py
from .types import SignEntry, SignType


def fingerspell(word: str) -> SignEntry:
    """Convert a word into a fingerspelled SignEntry (A-Z, 0-9)."""
    upper = word.strip().upper()
    letters = [c for c in upper if c.isalpha() or c.isdigit()]
    if not letters:
        letters = [upper]
    return SignEntry(
        gloss=upper,
        type=SignType.FINGERSPELL,
        duration_ms=len(letters) * 300,
        letters=letters,
    )
